qstash/workflows/howto/failures.mdx

```diff
@@ -0,0 +1,69 @@
+---
+title: "Handle Failures"
+---
+
+
+Failures can occur during a workflow run for various reasons:
+
+-  A step in your workflow throws an exception due to an error in your code. This could be anything from a database error to a logical bug that causes your code to fail during execution.
+-  QStash attempts to call your workflow URL, but the URL is not reachable. This might happen if your service is down resulting in a 404 error.
+-  Steps in your workflow might also fail if they exceed the [platform's time limitations](/qstash/workflows/platforms/platforms). When a step takes too long to execute, it can time out, causing the workflow to fail.
+
+In all of the cases above, QStash will retry calling the step 3 times and give up only after then. Making this configurable is in our [roadmap](qstash/workflows/roadmap)
+
+When a workflow failed, a related event will be fired to [Dead Letter Queue](/qstash/howto/handling-failures#dead-letter-queue)
+
+From `DLQ` tab on the console, you can check all the failed messages. You can filter messages via `workflow url` and `workflow run id` to search for a particilar failure.
+
+TOOD SANCAR PUT A DLQ IMAGE HERE WITH WORKFLOW URL WORKFLOW ID FILTERS
+
+If you want to take an action(a cleanup/log), you can configure either `failureFunction` or a `failureUrl` on the `serve` method of your workflow.
+These options allow you to define custom logic or an external endpoint that will be triggered when a failure occurs. 
+
+
+### 1. Using `failureFunction`
+
+The `failureFunction` allows you to define custom logic for handling failures. 
+
+#### Example Configuration:
+
+```javascript
+export const POST = serve<string>(
+  failureFunction: async (context, failStatus, failResponse) => {
+    const input = context.requestPayload;
+    cleanupDB(input)
+  },
```

Comment:
having the `context` parameter is really complex right now. It's partly made complex because of the auth-before-all-steps approach we have.

Here is the current parameters:
```
    status: number,
    header: Record<string, string>,
    body: FailureFunctionPayload,
    workflowRunId: string
```

Removing our auth approach, adding an onInvocation callback for authentication and adding context parameter to  failureFunction all combined is simpler than adding context to failureFunction in the current setup :D

TODO: parameter order is wrong too.